src/hooks/useSendEmail.ts

```diff
@@ -0,0 +1,78 @@
+import {
+  Dispatch,
+  SetStateAction,
+  useCallback,
+  useContext,
+} from 'react';
+import { useRouter } from 'next/router';
+import { useSearchParams } from 'next/navigation';
+import { AppContext } from '@context';
+import { decodeText } from '@utils';
+import { emailService } from '@services';
+import { useToast } from '@chakra-ui/react';
+
+type EmailServiceProps = {
+  onClose: () => void;
+  responderName: string;
+  setLoading: Dispatch<SetStateAction<boolean>>;
+};
+
+export const useSendEmail = () => {
```

Comment:
Okay, so I hate to be this butthead, but I really kinda don't like what you've done here. There's a lot of coupling and unclear encapsulation.

Specifically, this useSendEmail hook is controlling a LOT of behavior that has nothing to do with sending the email.

It is setting extrinsic state via setLoading, it is executing a close handler on a component via onClose, and it is rerouting a window. None of which, imo, should happen here. I'm also a little iffy on pulling the selected answer from the context, it really feels like that's some kind of constructed data that should be passed in, but I'll take it for now. Especially if you rename this to `useSendAnswerEmail`.

Let's take these items in order.

**setLoading**

This is an easy one. Why would a hook receive a setter for the state of its caller? What if its caller doesn't care about loading? What if its caller cares only about a more complicated, compound loading state? You've coupled this hook to its caller, and its caller to this hook's implementation. This is bad. The fix is easy. You just track loading state inside the hook and expose it in the return:

```
useSendEmail = () => {
  const [loading, setLoading] = useState(false);
  
  ...
  
  return { loading, sendEmail };
}
```

**onClose**

This is a hook. It isn't a modal. It's behavior, not a component. `onClose` is nonsense here, onClose what? The only reason you've injected an onClose here is because the caller needs it. That's a really bad reason to do something. What if the caller doesn't need an onClose? What if the caller needs some completely different action when the email is sent?  You have two options here. The first, and simplest, is to just create hooks that are named and conceptually bound to what's really happening: you can have this function take `onSuccess` and `onError` params, and call them in the `then` and `catch` respectively. Then pass your current onClose as onSuccess and call it day.

The second and cooler choice is just to expose enough state out of the hook that the caller can make all those decisions. Once again, I make reference to the Apollo Client useServer hooks (which I, frankly, think are near-perfect React API design). In this case, you're basically cloning Apollo Client's useLazyQuery API.

```
const useSendEmail = () => {
  const [loading, setLoading] = useState(false); // we've already discussed this
  const [error, setError] = useState(null); // this would get set to the contents of the error in your catch
  const [data, setData] = useState(); // this would get set to some data version of the final response. Maybe the current contents of the toast() call, and you call toast in the caller
  
  ...
  
  return { loading, data, error, sendEmail };
}
```

And then the caller can just monitor these values to perform its logic.

*Rerouting*

Actually, just the answer above handles re-routing as well.

*Selected Answer from Context*

This binds the hook to only being called from where the context is valid. This raises alarm bells, especially for a hook that should be as fundamental to an application as `useSendEmail`. Like, this feels like it should be an application-wide service, and should be able to send an email from anywhere in the application. Now, I get that this is a DYLM app, so a answer is pretty fundamental to the application's whole reason for existing. But it still feels like a mismatch in abstractions. Renaming it slightly would make me feel better, but so would just passing the data in.